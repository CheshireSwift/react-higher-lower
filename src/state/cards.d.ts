export type Suit = "H" | "S" | "C" | "D";
export type Rank = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13;
export interface Card {
  rank: Rank;
  suit: Suit;
}
