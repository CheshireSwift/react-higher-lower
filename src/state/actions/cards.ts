import { Card } from "../cards";

export const draw = (card: Card) => ({
  type: "DRAW" as "DRAW",
  payload: card
});

export const shuffle = () => ({
  type: "SHUFFLE" as "SHUFFLE"
});
