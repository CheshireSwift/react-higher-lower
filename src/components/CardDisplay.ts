import { div } from "react-hyperscript-helpers";

import { Card, Rank } from "../state/cards";

interface CardDisplayProps {
  card: Card | null;
  style?: any;
}

function formatRank(rank: Rank) {
  switch (rank) {
    case 11:
      return "J";
    case 12:
      return "Q";
    case 13:
      return "K";
    case 1:
      return "A";
    default:
      return rank;
  }
}

function formatCard(card: Card): string {
  return formatRank(card.rank) + card.suit;
}

export const CardDisplay = (props: CardDisplayProps) =>
  div(
    {
      style: {
        ...props.style,
        width: 40,
        height: 60,
        border: "1px solid black",
        background: props.card
          ? "white"
          : `radial-gradient(circle at 0% 50%, rgba(96, 16, 48, 0) 9px, #613 10px, rgba(96, 16, 48, 0) 11px) 0px 10px,
            radial-gradient(at 100% 100%,      rgba(96, 16, 48, 0) 9px, #613 10px, rgba(96, 16, 48, 0) 11px),
            #8a3`,
        backgroundSize: "20px 20px"
      }
    },
    props.card && formatCard(props.card)
  );

export default CardDisplay;
