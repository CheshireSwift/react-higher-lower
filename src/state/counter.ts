import { increment, decrement, set } from "./actions/tick";

interface CounterState {
  count: number;
}

export const counter: Reducer<
  CounterState,
  typeof increment | typeof decrement | typeof set
> = (state = { count: 0 }, action) => {
  switch (action.type) {
    case "INCREMENT":
      return { ...state, count: state.count + action.payload.amount };
    case "DECREMENT":
      return { ...state, count: state.count - action.payload.amount };
    case "SET":
      return { ...state, count: action.payload.value };
    default:
      return state;
  }
};
