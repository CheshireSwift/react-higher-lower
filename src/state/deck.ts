import _ from "lodash";

import { Card, Rank, Suit } from "./cards";
import { draw, shuffle } from "./actions/cards";

type DeckState = Card[];

const sortedDeck = (): DeckState =>
  _.flatten(
    (["S", "H", "C", "D"] as Suit[]).map(suit =>
      ([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13] as Rank[]).map(rank => ({
        suit,
        rank
      }))
    )
  );

export const deck: Reducer<DeckState, typeof draw | typeof shuffle> = (
  state = sortedDeck(),
  action
) => {
  switch (action.type) {
    case "DRAW":
      return _.reject(state, action.payload);
    case "SHUFFLE":
      return _.shuffle(state);
    default:
      return state;
  }
};
