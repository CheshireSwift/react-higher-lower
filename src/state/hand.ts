import _ from "lodash";

import { Card, Guess } from "./cards";
import { draw, guess } from "./actions/cards";

type HandState = Array<{ card: Card; guess: Guess | null }>;

export const hand: Reducer<HandState, typeof draw | typeof guess> = (
  state = [],
  action
) => {
  switch (action.type) {
    case "DRAW":
    case "GUESS":
      return [
        ...state,
        {
          card: action.payload.card,
          guess: _.get(action.payload, "direction", null)
        }
      ];
    default:
      return state;
  }
};
