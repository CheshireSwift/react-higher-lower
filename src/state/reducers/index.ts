import { combineReducers, Reducer } from "redux";

import { counter } from "./counter";

export type ReducerState<R> = R extends Reducer<infer S> ? S : null;
export const rootReducer = combineReducers({ counter });
