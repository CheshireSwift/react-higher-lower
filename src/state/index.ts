import { ReducerState, rootReducer } from "./reducers";
export type State = ReducerState<typeof rootReducer>;

export default State;
