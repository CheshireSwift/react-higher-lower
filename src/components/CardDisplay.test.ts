import { render } from "enzyme";
import { h } from "react-hyperscript-helpers";

import { Card } from "../state/cards";

import { CardDisplay } from "./CardDisplay";

describe("the Card Display", () => {
  const twoClubs: Card = { rank: 2, suit: "C" };

  const display = (card: Card | null) => render(h(CardDisplay, { card }));

  it("matches the back of card snapshot", () => {
    expect(display(null)).toMatchSnapshot();
  });

  it("matches the front of cardsnapshot", () => {
    expect(display(twoClubs)).toMatchSnapshot();
  });

  it("displays the card", () => {
    expect(display(twoClubs).text()).toContain("2C");
  });

  it("formats aces and faces", () => {
    function assertPair(card: Card, output: string) {
      expect(display(card).text()).toContain(output);
    }
    assertPair({ rank: 1, suit: "S" }, "AS");
    assertPair({ rank: 11, suit: "C" }, "JC");
    assertPair({ rank: 12, suit: "H" }, "QH");
    assertPair({ rank: 13, suit: "D" }, "KD");
  });
});
