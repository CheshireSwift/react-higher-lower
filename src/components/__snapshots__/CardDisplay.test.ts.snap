// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Card Display matches the back of card snapshot 1`] = `
<div
  style="width:40px;height:60px;border:1px solid black;background:radial-gradient(circle at 0% 50%, rgba(96, 16, 48, 0) 9px, #613 10px, rgba(96, 16, 48, 0) 11px) 0px 10px,
            radial-gradient(at 100% 100%,      rgba(96, 16, 48, 0) 9px, #613 10px, rgba(96, 16, 48, 0) 11px),
            #8a3;background-size:20px 20px"
/>
`;

exports[`the Card Display matches the front of cardsnapshot 1`] = `
<div
  style="width:40px;height:60px;border:1px solid black;background:white;background-size:20px 20px"
>
  2C
</div>
`;
